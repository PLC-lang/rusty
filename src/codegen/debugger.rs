
use inkwell::debug_info::{DebugInfoBuilder,  DICompileUnit, DIType, DIScope, DISubprogram, AsDIScope};
use inkwell::context::Context;
use inkwell::builder::Builder;

use crate::ast::*;
use crate::index::DataTypeInformation;

pub struct DebugManager<'ctx> {

    context: Option<&'ctx Context>,
    di_builder: Option<DebugInfoBuilder<'ctx>>,
    compilation_unit: Option<DICompileUnit<'ctx>>,
    scope:  Option<DIScope<'ctx>>,
}

impl <'ctx> DebugManager<'ctx> {
    pub fn create_inactive() -> DebugManager<'ctx> {
        DebugManager::new(None,None, None)
    }

    pub fn activate(&mut self, context: &'ctx Context, di_builder : DebugInfoBuilder<'ctx>, compilation_unit: DICompileUnit<'ctx>) {
        self.context = Some(context);
        self.di_builder = Some(di_builder);
        self.compilation_unit = Some(compilation_unit);
    }
    
    pub fn new(context: Option<&'ctx Context>,  di_builder : Option<DebugInfoBuilder<'ctx>>, compilation_unit: Option<DICompileUnit<'ctx>>) -> DebugManager<'ctx> {
        DebugManager {
            context,
            di_builder,
            compilation_unit,
            scope: None,
        }
    }

    pub fn finalize(&self) {
        if let Some(di_builder) =  &self.di_builder {
            di_builder.finalize();       
        }
    }

    pub fn create_subprogram(&mut self, members : &[(&str, DataTypeInformation<'ctx>)], function : &POU, new_lines : &[usize]) -> Option<(DIType<'ctx>, DISubprogram<'ctx>)>{
        //Exit if dummy
        if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();

        let line_number = get_line_of(new_lines, &function.location.start) as u32 +1 ; //TODO: Line numbers are off by one
        //
        //Create and get a struct type
        let struct_type = self.get_struct_type(function.name.as_str(), line_number, members).unwrap();

        //Add function debug information
        let scope = self.get_current_scope();
        let subroutine_type = di_builder.create_subroutine_type(compilation_unit.get_file(), 
        None,  &[struct_type] , inkwell::debug_info::DIFlagsConstants::PUBLIC, );
        let result = di_builder.create_function(
            scope, 
            function.name.as_str(), 
            None, 
            compilation_unit.get_file(), 
            line_number, 
            subroutine_type, 
            true, 
            true, 
            line_number, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC, 
            false
        );
        self.scope = Some(result.as_debug_info_scope());
        Some((struct_type, result))
    }

    fn get_current_scope(&self) -> DIScope<'ctx> {
        self.scope.or(Some(self.compilation_unit.as_ref().unwrap().get_file().as_debug_info_scope())).unwrap()
    }

    pub fn get_basic_type(&self, name : &str, datatype: &DataTypeInformation) -> Option<DIType<'ctx>> {
         //Exit if dummy
         if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        //TODO : Cache
        match *datatype {
            DataTypeInformation::Integer {size, ..} => {
                di_builder.create_basic_type(
                    name, 
                    size as u64, 
                    0x00, 
                    inkwell::debug_info::DIFlagsConstants::PUBLIC
                ).map(|it| it.as_type()).ok()
            },
            _ => None,
        }   
    }
/*
    pub fn get_instance_variable(&self, name : &str, debug_type : DIType<'ctx>, line : u32,global_value : PointerValue,block : BasicBlock, debug_location : DILocation) {
        //Exit if dummy
        if self.di_builder.is_none() {
            return;
        }

        let context = self.context.unwrap();
        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();
        
        let instance_variable = Some(di_builder.create_parameter_variable(
            self.get_current_scope(), 
            name, 
            0, 
            compilation_unit.get_file(), 
            line, 
            debug_type, 
            false, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC,
        ));

        let debug_loc = di_builder.create_debug_location(context, line, 0, self.get_current_scope(), None);
        di_builder.insert_declare_at_end(
            global_value,
             instance_variable, 
             Some(di_builder.create_expression(Vec::new())), 
             debug_loc, 
             block
        );
    }
*/
    /* fn get_variables_debug_information(
        &self,
        debugger: &mut DebugManager<'ctx>,
        variables: &Vec<Variable>,
    ) -> Vec<(String, Option<DIType<'ctx>>)>
    {
        let mut types: Vec<(String, Option<DIType<'ctx>>)> = Vec::new();
        for variable in variables {
            let (_, debug_type)  = self.get_debug_type(debugger, &variable.data_type).unwrap();
            types.push((variable.name.clone(), debug_type));
        }
        types
    }

    fn get_debug_type(&self, name: &str, data_type: &DataTypeInformation<'ctx>) -> Option<(BasicTypeEnum<'ctx>, Option<DIType<'ctx>>)> {
        Some(data_type.get_type(), self.get_basic_type(name, data_type))
    }
*/
    pub fn get_struct_type(&self, name : &str, line : u32, members : &[(&str, DataTypeInformation<'ctx>)]) -> Option<DIType<'ctx>> {
        //Exit if dummy
        if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();


        let mut elements = Vec::new();
        let mut size = 0;
        for (name, datatype) in members {
            size = size + datatype.get_size();
            if let Some(member) = self.get_basic_type(name, datatype) {
                elements.push(member);
            }
        }

        let result = di_builder.create_struct_type(
            self.get_current_scope(), 
            name, 
            compilation_unit.get_file(), 
            line, 
            size as u64, 
            8, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC, 
            None, 
            elements.as_slice(), 
            0, 
            None, 
            format!("{}_debug_interface", name.to_string()).as_str(),
        );

        di_builder.create_typedef(result.as_type(), name, compilation_unit.get_file(), line
            , self.get_current_scope(), 8);


        let mut current_offset = 0;
        for (i, (name, datatype)) in members.iter().enumerate() {
            di_builder.create_member_type(result.as_debug_info_scope(), name, compilation_unit.get_file(), line, (datatype.get_size() * 8) as u64, 8, current_offset, 
                inkwell::debug_info::DIFlagsConstants::PUBLIC, 
                elements[i],
            );
            current_offset = current_offset + (datatype.get_size() * 8) as u64; //TODO : Add align magic
        }

        Some(result.as_type())
    }

    pub fn append_breakpoint_location(&mut self, builder : &Builder, line : u32) {
        //Exit if dummy
        if self.di_builder.is_none() {
            return;
        }
        let context = self.context.unwrap();
        let di_builder = self.di_builder.as_ref().unwrap();
        
        let parent_scope = self.get_current_scope();
        //let compilation_unit = self.compilation_unit.as_ref().unwrap();
        //let lexical_block = di_builder.create_lexical_block(parent_scope, compilation_unit.get_file(), line, 0);
        
        let result = di_builder.create_debug_location(context, line, 0, parent_scope, None);
        builder.set_current_debug_location(context, result)

    }

    pub fn remove_scope(&mut self) {
        self.scope = None;
    }
}
