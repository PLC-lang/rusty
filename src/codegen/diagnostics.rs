use std::{backtrace, fmt::Display};

use inkwell::builder::BuilderError;
use plc_diagnostics::diagnostics::Diagnostic;
use plc_source::source_location::SourceLocation;
use thiserror::Error;

#[derive(Error, Debug)]
pub struct CodegenDiagnostic {
    message: String,
    source: BuilderError,
    location: SourceLocation,
}

impl Display for CodegenDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Codegen error: {}. Source: {} at {}", &self.message, &self.source, &self.location)
    }
}

impl From<BuilderError> for CodegenDiagnostic {
    fn from(source: BuilderError) -> Self {
        // println!("Backtrace: {} ", backtrace::Backtrace::force_capture());
        CodegenDiagnostic { message: source.to_string(), source, location: SourceLocation::undefined() }
    }
}

impl CodegenDiagnostic {
    pub(crate) fn with_location(self, location: SourceLocation) -> CodegenDiagnostic {
        CodegenDiagnostic { location, ..self }
    }
}

impl From<CodegenDiagnostic> for Diagnostic {
    fn from(value: CodegenDiagnostic) -> Self {
        Diagnostic::codegen_error(&value.message, value.location.clone()).with_internal_error(value.into())
    }
}
