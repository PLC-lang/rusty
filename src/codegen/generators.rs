pub mod data_type_generator;
pub mod date_time_util;
pub mod expression_generator;
pub mod llvm;
pub mod pou_generator;
pub mod statement_generator;
pub mod variable_generator;

// see: https://llvm.org/doxygen/NVPTXBaseInfo_8h_source.html
pub const ADDRESS_SPACE_GENERIC: u16 = 0;
pub const ADDRESS_SPACE_GLOBAL: u16 = 1;
pub const ADDRESS_SPACE_CONST: u16 = 4;
