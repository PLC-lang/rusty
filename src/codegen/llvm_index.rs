// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
use crate::{ast::SourceRange, compile_error::CompileError};
use inkwell::types::BasicTypeEnum;
use inkwell::values::{BasicValueEnum, FunctionValue, GlobalValue, PointerValue};
use std::collections::HashMap;

/// Index view containing declared values for the current context
/// Parent Index is the a fallback lookup index for values not declared locally
#[derive(Debug, Clone)]
pub struct LlvmTypedIndex<'ink> {
    parent_index: Option<&'ink LlvmTypedIndex<'ink>>,
    type_associations: HashMap<String, BasicTypeEnum<'ink>>,
    initial_value_associations: HashMap<String, BasicValueEnum<'ink>>,
    loaded_variable_associations: HashMap<String, PointerValue<'ink>>,
    implementations: HashMap<String, FunctionValue<'ink>>,
    constants: HashMap<String, BasicValueEnum<'ink>>,
}

impl<'ink> LlvmTypedIndex<'ink> {
    pub fn new() -> LlvmTypedIndex<'ink> {
        LlvmTypedIndex {
            parent_index: None,
            type_associations: HashMap::new(),
            initial_value_associations: HashMap::new(),
            loaded_variable_associations: HashMap::new(),
            implementations: HashMap::new(),
            constants: HashMap::new(),
        }
    }

    pub fn create_child(parent: &'ink LlvmTypedIndex<'ink>) -> LlvmTypedIndex<'ink> {
        LlvmTypedIndex {
            parent_index: Some(parent),
            type_associations: HashMap::new(),
            initial_value_associations: HashMap::new(),
            loaded_variable_associations: HashMap::new(),
            implementations: HashMap::new(),
            constants: HashMap::new(),
        }
    }

    pub fn merge(&mut self, mut other: LlvmTypedIndex<'ink>) {
        for (name, assocication) in other.type_associations.drain() {
            self.type_associations.insert(name, assocication);
        }
        for (name, assocication) in other.initial_value_associations.drain() {
            self.initial_value_associations.insert(name, assocication);
        }
        for (name, assocication) in other.initial_value_associations.drain() {
            self.initial_value_associations.insert(name, assocication);
        }
        for (name, assocication) in other.loaded_variable_associations.drain() {
            self.loaded_variable_associations.insert(name, assocication);
        }
        for (name, implementation) in other.implementations.drain() {
            self.implementations.insert(name, implementation);
        }
        self.constants.extend(other.constants);
    }

    pub fn associate_type(
        &mut self,
        type_name: &str,
        target_type: BasicTypeEnum<'ink>,
    ) -> Result<(), CompileError> {
        self.type_associations
            .insert(type_name.to_lowercase(), target_type);
        Ok(())
    }

    pub fn associate_initial_value(
        &mut self,
        type_name: &str,
        initial_value: BasicValueEnum<'ink>,
    ) -> Result<(), CompileError> {
        self.initial_value_associations
            .insert(type_name.to_lowercase(), initial_value);
        Ok(())
    }

    pub fn associate_loaded_local_variable(
        &mut self,
        container_name: &str,
        variable_name: &str,
        target_value: PointerValue<'ink>,
    ) -> Result<(), CompileError> {
        let qualified_name = format!("{}.{}", container_name, variable_name);
        self.loaded_variable_associations
            .insert(qualified_name.to_lowercase(), target_value);
        Ok(())
    }

    pub fn find_associated_type(&self, type_name: &str) -> Option<BasicTypeEnum<'ink>> {
        self.type_associations
            .get(&type_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_type(type_name))
                    .flatten()
            })
    }

    pub fn get_associated_type(
        &self,
        type_name: &str,
    ) -> Result<BasicTypeEnum<'ink>, CompileError> {
        self.find_associated_type(type_name)
            .ok_or_else(|| CompileError::unknown_type(type_name, SourceRange::undefined()))
    }

    pub fn find_associated_initial_value(&self, type_name: &str) -> Option<BasicValueEnum<'ink>> {
        self.initial_value_associations
            .get(&type_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_initial_value(type_name))
                    .flatten()
            })
    }

    pub fn associate_global(
        &mut self,
        variable_name: &str,
        global_variable: GlobalValue<'ink>,
    ) -> Result<(), CompileError> {
        self.initial_value_associations.insert(
            variable_name.to_lowercase(),
            global_variable.as_pointer_value().into(),
        );
        Ok(())
    }

    pub fn associate_implementation(
        &mut self,
        callable_name: &str,
        function_value: FunctionValue<'ink>,
    ) -> Result<(), CompileError> {
        self.implementations
            .insert(callable_name.to_lowercase(), function_value);
        Ok(())
    }

    pub fn find_associated_implementation(
        &self,
        callable_name: &str,
    ) -> Option<FunctionValue<'ink>> {
        self.implementations
            .get(&callable_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_implementation(callable_name))
                    .flatten()
            })
    }

    pub fn find_associated_variable_value(
        &self,
        qualified_name: &str,
    ) -> Option<BasicValueEnum<'ink>> {
        self.initial_value_associations
            .get(&qualified_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_variable_value(qualified_name))
                    .flatten()
            })
    }

    pub fn find_loaded_associated_variable_value(
        &self,
        qualified_name: &str,
    ) -> Option<PointerValue<'ink>> {
        let result = self
            .loaded_variable_associations
            .get(&qualified_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_loaded_associated_variable_value(qualified_name))
                    .flatten()
            });

        //If nothing got associated, see if we have a global we could reuse
        result.or_else(|| {
            self.find_associated_variable_value(qualified_name)
                .filter(|it| it.is_pointer_value())
                .map(BasicValueEnum::into_pointer_value)
        })
    }

    pub fn associate_constant(
        &mut self,
        qualified_name: &str,
        basic_value_enum: BasicValueEnum<'ink>,
    ) {
        self.constants
            .insert(qualified_name.into(), basic_value_enum);
    }

    pub fn find_constant_value(&self, qualified_name: &str) -> Option<BasicValueEnum<'ink>> {
        self.constants.get(qualified_name).copied()
    }
}
