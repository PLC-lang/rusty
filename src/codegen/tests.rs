// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod code_gen_tests;
mod codegen_error_messages_tests;
mod compare_instructions_tests;
mod constants_tests;
mod debug_tests;
mod directaccess_test;
mod expression_tests;
mod function_tests;
mod generics_test;
mod initialization_test;
mod parameters_tests;
mod statement_codegen_test;
mod string_tests;
#[cfg(feature = "verify")]
mod switch_case_tests;
mod typesystem_test;
