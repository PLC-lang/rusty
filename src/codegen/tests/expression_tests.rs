// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
use crate::test_utils::tests::codegen;

#[test]
fn pointers_in_function_return() {
    let result = codegen(
        r#"FUNCTION func : REF_TO INT
        END_FUNCTION"#,
    );
    insta::assert_snapshot!(result);
}

#[test]
fn structs_in_function_return() {
    let result = codegen(
        r#"
        TYPE myStruct : STRUCT
            x : INT;
            END_STRUCT
        END_TYPE
        FUNCTION func : myStruct
            VAR_OUTPUT
                xxx : myStruct;
            END_VAR
        END_FUNCTION"#,
    );
    insta::assert_snapshot!(result);
}

#[test]
fn strings_in_function_return() {
    let result = codegen(
        r#"
       FUNCTION func : STRING
            VAR_INPUT
                myout : REF_TO STRING;
            END_VAR

            func := 'hello';
            myout^ := 'hello';
       END_FUNCTION"#,
    );
    insta::assert_snapshot!(result);
}

#[test]
fn calling_strings_in_function_return() {
    let result = codegen(
        r#"
       FUNCTION func : STRING
            func := 'hello';
       END_FUNCTION

       PROGRAM main
            VAR
                x : STRING;
            END_VAR

            x := func();
       END_PROGRAM
       "#,
    );
    insta::assert_snapshot!(result);
}

#[test]
fn unary_expressions_can_be_real() {
    let result = codegen(
        r#"
            PROGRAM prg
            VAR
                a,b : REAL;
            END_VAR
                b := -2.0;
                a := -b;
            END_PROGRAM
        "#,
    );
    insta::assert_snapshot!(result);
}

#[test]
fn type_mix_in_call() {
    let result = codegen(
        "
        FUNCTION foo : INT
        VAR_INPUT
            in : INT;
        END_VAR
        END_FUNCTION
        FUNCTION baz : INT
            foo(1.5);
        END_FUNCTION
    ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn cast_pointer_to_lword() {
    let result = codegen(
        r#"
        FUNCTION baz : INT
            VAR
                ptr_x : POINTER TO INT;
                y : LWORD;
            END_VAR;

            y := ptr_x;
        END_FUNCTION
    "#,
    );

    //should result in normal number-comparisons
    insta::assert_snapshot!(result);
}

#[test]
fn cast_lword_to_pointer() {
    let result = codegen(
        r#"
        FUNCTION baz : INT
            VAR
                ptr_x : POINTER TO INT;
                y : LWORD;
            END_VAR;

            ptr_x := y;
        END_FUNCTION
    "#,
    );

    //should result in normal number-comparisons
    insta::assert_snapshot!(result);
}

#[test]
fn cast_between_pointer_types() {
    let result = codegen(
        r#"
        PROGRAM baz
            VAR
                ptr_x : POINTER TO BYTE;
                y : WORD;
            END_VAR;

            ptr_x := REF(y);
        END_PROGRAM
    "#,
    );

    //should result in bitcast conversion when assigning to ptr_x
    insta::assert_snapshot!(result);
}

#[test]
fn unnecessary_casts_between_pointer_types() {
    let result = codegen(
        r#"
        TYPE MyByte : BYTE; END_TYPE

        PROGRAM baz
            VAR
                ptr : POINTER TO BYTE;
                b : BYTE;
                si : SINT;
                mb : MyByte;
            END_VAR;

            ptr := REF(b); //no cast necessary
            ptr := REF(si); //no cast necessary
            ptr := REF(mb); //no cast necessary
        END_PROGRAM
    "#,
    );

    //should not result in bitcast
    insta::assert_snapshot!(result);
}

#[test]
fn access_string_via_byte_array() {
    let result = codegen(
        r#"
        TYPE MyByte : BYTE; END_TYPE

        PROGRAM baz
            VAR
                str: STRING[10];
                ptr : POINTER TO BYTE;
                bytes : POINTER TO ARRAY[0..9] OF BYTE;
            END_VAR;

            ptr := REF(str); //bit-cast expected
            bytes := REF(str);
        END_PROGRAM
    "#,
    );

    //should result in bitcasts
    insta::assert_snapshot!(result);
}

#[test]
fn pointer_arithmetics() {
    // codegen should be successful for binary expression for pointer<->int / int<->pointer / pointer<->pointer
    let result = codegen(
        "
        PROGRAM main
        VAR
            x : INT := 10;
            y : INT := 20;
            pt : REF_TO INT;
        END_VAR
        pt := REF(x);

        (* +/- *)
        pt := pt + 1;
        pt := pt + 1 + 1;
        pt := 1 + pt;
        pt := pt - y;
        pt := 1 + pt + 1;
        pt := pt - y - 1;
        pt := 1 + 1 + pt ;
        pt := y + pt - y ;
        pt := y + y + pt ;
        END_PROGRAM
        ",
    );
    insta::assert_snapshot!(result);
}

#[test]
fn pointer_arithmetics_function_call() {
    // codegen should be successful for binary expression for pointer<->int / int<->pointer / pointer<->pointer
    let result = codegen(
        "
        FUNCTION foo : LINT
        END_FUNCTION

        PROGRAM main
        VAR
            pt : REF_TO INT;
            x : INT;
        END_VAR
        pt := REF(x);

        (* +/- *)
        pt := pt + foo();
        END_PROGRAM
        ",
    );
    insta::assert_snapshot!(result);
}

#[test]
fn nested_call_statements() {
    // GIVEN some nested call statements
    let result = codegen(
        "
        FUNCTION foo : DINT
        VAR_INPUT
            a : DINT;
        END_VAR
        END_FUNCTION

        PROGRAM main
            foo(foo(2));
        END_PROGRAM
        ",
    );
    // WHEN compiled
    // WE expect a flat sequence of calls, no regions and branching
    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_adr() {
    // GIVEN some nested call statements
    let result = codegen(
        "
        PROGRAM main
        VAR
            a : REF_TO DINT;
            b : DINT;
        END_VAR
            a := ADR(b);
        END_PROGRAM
        ",
    );
    // WHEN compiled
    // We expect a direct conversion to lword and subsequent assignment (no call)
    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_ref() {
    // GIVEN some nested call statements
    let result = codegen(
        "
        PROGRAM main
        VAR
            a : REF_TO DINT;
            b : DINT;
        END_VAR
            a := REF(b);
        END_PROGRAM
        ",
    );
    // WHEN compiled
    // We expect a direct conversion and subsequent assignment to pointer(no call)
    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_mux() {
    let result = codegen(
        "PROGRAM main
        VAR
            a,b,c,d,e : DINT;
        END_VAR
            a := MUX(3, b,c,d,e); // 3 = e
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_mux_with_aggregate_type() {
    let result = codegen(
        "PROGRAM main
        VAR
            str1 : STRING;
        END_VAR
            str1 := MUX(3, 'lorem', 'ipsum', 'dolor', 'sit'); // 3 = sit
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_sel() {
    let result = codegen(
        "PROGRAM main
        VAR
            a,b,c : DINT;
        END_VAR
            a := SEL(TRUE, b,c);
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_sel_as_expression() {
    let result = codegen(
        "PROGRAM main
        VAR
            a,b,c : DINT;
        END_VAR
            a := SEL(TRUE, b,c) + 10;
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_move() {
    let result = codegen(
        "PROGRAM main
        VAR
            a,b : DINT;
        END_VAR
            a := MOVE(b);
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_sizeof() {
    let result = codegen(
        "PROGRAM main
        VAR
            a: DINT;
            b: LINT;
        END_VAR
            a := SIZEOF(b);
        END_PROGRAM",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_lower_bound() {
    let result = codegen(
        "PROGRAM main
        VAR
            a: ARRAY[0..1] OF DINT;
            b: DINT;
        END_VAR
            b := foo(a);
        END_PROGRAM

        FUNCTION foo : DINT
        VAR_IN_OUT
            vla: ARRAY[*] OF DINT;
        END_VAR
            foo := LOWER_BOUND(vla, 1);
        END_VAR
        END_FUNCTION
        ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_upper_bound() {
    let result = codegen(
        "PROGRAM main
        VAR
            a: ARRAY[0..1] OF DINT;
            b: DINT;
        END_VAR
            b := foo(a);
        END_PROGRAM

        FUNCTION foo : DINT
        VAR_IN_OUT
            vla: ARRAY[*] OF DINT;
        END_VAR
            foo := UPPER_BOUND(vla, 1);
        END_VAR
        END_FUNCTION
        ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_function_call_upper_bound_expr() {
    let result = codegen(
        "
        VAR_GLOBAL CONSTANT
            MY_CONST : DINT := 10;
        END_VAR
        PROGRAM main
        VAR
            a: ARRAY[0..1, 1..2, 2..3, 3..4] OF DINT;
            b: DINT;
        END_VAR
            b := foo(a);
        END_PROGRAM

        FUNCTION foo : DINT
        VAR_IN_OUT
            vla: ARRAY[*] OF DINT;
        END_VAR
            // upper bound of 4th dimension => 8th element in dimension array
            foo := UPPER_BOUND(vla, MY_CONST - (2 * 3));
        END_VAR
        END_FUNCTION
        ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn test_max_int() {
    let result = codegen(
        r"
    {external}
    FUNCTION MAX<U : ANY> : U
    VAR_INPUT in : {sized} U...; END_VAR
    END_FUNCTION

    FUNCTION main : INT
    main := MAX(INT#5,INT#2,INT#1,INT#3,INT#4,INT#7,INT#-1);
    END_FUNCTION",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn compare_date_time_literals() {
    let result = codegen(
        "
    PROGRAM main
    VAR_TEMP
        cmp1, cmp2, cmp3, cmp4, cmp5, cmp6, cmp7, cmp8 : BOOL;
    END_VAR
        cmp1 := TIME#2d4h6m8s10ms11us300ns < TIME#1d8h43m23s55ms;
        cmp2 := LTIME#2d4h6m8s10ms11us300ns > LTIME#1d8h43m23s55ms;

        cmp3 := TOD#23:59:59.999 < TOD#10:32:59;
        cmp4 := LTOD#23:59:59.999 > LTOD#10:32:59;

        cmp5 := DATE#2022-10-20 < DATE#1999-01-01;
        cmp6 := LDATE#2022-10-20 > LDATE#1999-01-01;

        cmp7 := DT#2022-10-20-23:59:59.999 < DT#1999-01-01-10:32;
        cmp8 := LDT#2022-10-20-23:59:59.999 > LDT#1999-01-01-10:32;
    END_PROGRAM
    ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn hardware_access_codegen() {
    let result = codegen(
        "
        PROGRAM prg
        VAR
          x,y,z : BYTE;
        END_VAR
          x := %IB1.2;
          y := %MB1.2;
          z := %GB1.2;
          x := %IX1.2;
          y := %MD1.2;
          z := %GW1.2;
        END_PROGRAM
        ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn hardware_access_assign_codegen() {
    let result = codegen(
        "
        PROGRAM prg
        VAR
          x,y,z : BYTE;
        END_VAR
          %IB1.2 := 1;
          %MB1.2 := 1;
          %GB1.2 := 1;
          %IX1.2 := 1;
          %MD1.2 := 1;
          %GW1.2 := 1;
        END_PROGRAM
        ",
    );

    insta::assert_snapshot!(result);
}

#[test]
fn allowed_assignable_types() {
    let result = codegen(
        r#"
        PROGRAM main
        VAR
            v : INT;
            x : ARRAY[0..1] OF INT;
            y : REF_TO INT;
            z : REF_TO ARRAY[0..1] OF INT;
        END_VAR
            v := 0;
            x[0] := 1;
            y^ := 2;
            y^.1 := 3;
            z^[0] := 4;
            z^[1].1 := 5;
        END_PROGRAM
        "#,
    );

    insta::assert_snapshot!(result);
}

#[test]
fn builtin_add_ints() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : DINT;
            l1 : LINT;
            s1 : SINT;
        END_VAR
            ADD(x1, x2, x3, l1, s1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_add_float() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : REAL;
            l1 : LREAL;
        END_VAR
            ADD(x1, x2, x3, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_add_mixed() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : REAL;
            l1 : LINT;
        END_VAR
            ADD(x1, x2, x3, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_mul_ints() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : DINT;
            l1 : LINT;
            s1 : SINT;
        END_VAR
            MUL(x1, x2, x3, l1, s1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_mul_float() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : REAL;
            l1 : LREAL;
        END_VAR
            MUL(x1, x2, x3, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_mul_mixed() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1, x2, x3 : REAL;
            l1 : LINT;
        END_VAR
            MUL(x1, x2, x3, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_sub_ints() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : DINT;
            l1 : LINT;
        END_VAR
            SUB(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_sub_float() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : REAL;
            l1 : LREAL;
        END_VAR
            SUB(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_sub_mixed() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : REAL;
            l1 : LINT;
        END_VAR
            SUB(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_div_ints() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : DINT;
            l1 : LINT;
        END_VAR
            DIV(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_div_float() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : REAL;
            l1 : LREAL;
        END_VAR
            DIV(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}

#[test]
fn builtin_div_mixed() {
    let src = r#"
        FUNCTION main : DINT
        VAR
            x1 : REAL;
            l1 : LINT;
        END_VAR
            DIV(x1, l1);
        END_FUNCTION
    "#;

    let res = codegen(src);

    insta::assert_snapshot!(res);
}
