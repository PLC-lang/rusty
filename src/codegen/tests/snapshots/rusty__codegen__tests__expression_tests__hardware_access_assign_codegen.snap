---
source: src/codegen/tests/expression_tests.rs
expression: result
---
; ModuleID = 'main'
source_filename = "main"

%prg = type { i8, i8, i8 }

@prg_instance = global %prg zeroinitializer

define void @prg(%prg* %0) {
entry:
  %x = getelementptr inbounds %prg, %prg* %0, i32 0, i32 0
  %y = getelementptr inbounds %prg, %prg* %0, i32 0, i32 1
  %z = getelementptr inbounds %prg, %prg* %0, i32 0, i32 2
  ret void
}

